internal/check/engine.go

```diff
@@ -0,0 +1,87 @@
+package check
+
+import (
+	"context"
+	"errors"
+
+	"github.com/ory/herodot"
+
+	"github.com/ory/keto/internal/relationtuple"
+	"github.com/ory/keto/internal/x"
+)
+
+type (
+	EngineProvider interface {
+		PermissionEngine() *Engine
+	}
+	Engine struct {
+		d EngineDependencies
+	}
+	EngineDependencies interface {
+		relationtuple.ManagerProvider
+	}
+)
+
+func NewEngine(d EngineDependencies) *Engine {
+	return &Engine{
+		d: d,
+	}
+}
+
+func (e *Engine) subjectIsAllowed(ctx context.Context, requested *relationtuple.InternalRelationTuple, rels []*relationtuple.InternalRelationTuple) (bool, error) {
+	// This is the same as the graph problem "can requested.Subject be reached from requested.Object through the first outgoing edge requested.Relation"
+	//
+	// recursive breadth-first search
+	// TODO replace by more performant algorithm
+
+	var allowed bool
+	for _, sr := range rels {
+		// we only have to check Subject here as we know that sr was reached from requested.ObjectID, requested.Relation through 0...n indirections
+		if requested.Subject.Equals(sr.Subject) {
+			// found the requested relation
+			return true, nil
+		}
+
+		sub, isSubjectSet := sr.Subject.(*relationtuple.SubjectSet)
+		if !isSubjectSet {
+			continue
+		}
+
+		var err error
+		// expand the set by one indirection; paginated
+		allowed, err = e.checkOneIndirectionFurther(ctx, requested, &relationtuple.RelationQuery{Object: sub.Object, Relation: sub.Relation, Namespace: sub.Namespace})
+		if err != nil {
+			return false, err
+		}
+		if allowed {
+			return true, nil
+		}
+	}
+
+	return allowed, nil
```

Comment:
The default cause is false, not a variable. Remember the macOS TLS vulnerability? ;)

```suggestion
	return false, nil
```

Reply:
And probably this is caused by the same issue as the macOS vuln: reminiscence of refactoring